services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java

```diff
@@ -163,7 +164,36 @@ private SideEffect handleSpendRequest(BigInteger promotionId, UUID basketId, UUI
         // using tid as user choice TODO change this once user choice generation is properly implemented, see issue 75
         DeductOutput spendProviderOutput = incentiveSystem.generateSpendRequestResponse(promotion.getPromotionParameters(), spendRequest, new ProviderKeyPair(providerSecretKey, providerPublicKey), tid, spendDeductTree, tid);
 
-        offlineDspRepository.addToDbSyncQueue(promotionId, tid, spendRequest, spendProviderOutput);
+        /*
+        * Incentive service queries double-spending protection service for whether dsid of spent token is already contained.
+        * If yes: abort transaction (since trivially identified as double-spending) and return dedicated caught-double-spending side effect.
+        * Else: spend request is answered as normal, transaction is recorded into the database as soon as possible
+        *
+        * If dsp service is down at the time of the above query, the transaction is recorded in the database regardless of the used dsid.
+        *
+        * Note that the above check is one-sided, i.e. while it never wrongly identifies a transaction as double-spending,
+        * it cannot identify all invalid transactions.
+        * An example scenario where the check does not notice an invalid transaction would be the following:
+        * 1. Customer spends token t1 in store A, obtaining remainder token t1'.
+        * 2. Store B is temporarily disconnected from the double-spending protection service
+        * 3. Customer spends token t1 in store B, obtaining remainder token t1''.
+        * 4. Customer spends token t1'' in store C.
+        * 5. Store B reconnects to the dsprotection service and syncs transaction into database.
+        *    => it was not known to dsprotection service (thus also to the incentive service) in step 4 that the spent token resulted from a double-spending transaction
+        *    => transaction was recorded in step 4 despite it is invalid
+        *    => needs to be invalidated now
+        *
+        * For the invalidation in step 5 to be possible,
+        * it is critical that transactions that occured during dsp service downtime are always recorded in the database,
+        * no matter whether dsid was already known.
+        */
+        GroupElement usedTokenDsid = spendRequest.getDsid();
+        if(offlineDspRepository.dspServiceIsAlive() && offlineDspRepository.containsDsid(usedTokenDsid)) {
+            return new CaughtDoubleSpendingSideEffect("Double-spending attempt detected: " + usedTokenDsid + " has already been spent!");
+        }
```

Comment:
The `offlineDspRepository.dspServiceIsAlive()` only checks that the server is not physically down, right? (and it seems like it would throw an unfriendly exception instead of returning `false`).
This should probably be replaced by a check whether or not the store has pretend-lost connection to the database 🤔.

Reply:
Maybe just let the containsDsid call throw an offline exception triggered by the simulated offline feature

Reply:
What about this? Did you build in Paul's recommendation?

Reply:
Why would it throw an unfriendly exception? If the blocking call in DsProtectionClient.dspIsAlive() in client package times out, the returned String is null as far as I know. So the method will return false.

Reply:
[the documentation](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#block-java.time.Duration-) reads to me like it would raise a `RuntimeException`. Maybe I'm missing something there.
([Context](https://github.com/cryptimeleon/incentive-system/blob/424aadfe80c914e74baaa9d8ceaedf69fed0ffaa/services/client/src/main/java/org/cryptimeleon/incentive/client/DSProtectionClient.java#L123)).

But in any case, you should make it so that if we're "simulated offline", then you shouldn't check the dsid.